src/tactic/move_add.lean

```diff
@@ -0,0 +1,313 @@
+  end
+
+/-- Partially traverses an expression in search for a sum of terms.
+When `recurse_on_expr` finds a sum, it sorts it using `sorted_sum`. -/
+meta def recurse_on_expr (hyp : option name) (ll : list (bool × pexpr)) : expr → tactic (list bool)
+| e@`(%%_ + %%_)     := sorted_sum hyp ll e
+| (expr.lam _ _ _ e) := recurse_on_expr e
+| (expr.pi  _ _ _ e) := recurse_on_expr e
+| e                  := do
+  li_unused ← e.get_app_args.mmap recurse_on_expr,
+  let unused_summed := li_unused.transpose.map list.band,
+  return unused_summed
+
+/-- Passes the user input `ll` to `recurse_on_expr` at a single location, that could either be
+`none` (referring to the goal) or `some name` (referring to hypothesis `name`).  Returns a pair
+consisting of a boolean and a further list of booleans.  The single boolean is `tt` if the tactic
+did *not* change the goal on which it was acting.  The list of booleans records which variable in
+`ll` has been unified in the application: `tt` means that the corresponding variable has *not* been
+unified.
+
+This definition is useful to streamline error catching. -/
+meta def move_add_with_errors (ll : list (bool × pexpr)) : option name → tactic (bool × list bool)
+| (some hyp) := do
+  lhyp ← get_local hyp,
+  thyp ←  infer_type lhyp,
+  is_unused ← recurse_on_expr hyp ll thyp,
+  nhyp ← get_local hyp,
+  nthyp ← infer_type nhyp,
+  if (thyp = nthyp) then return (tt, is_unused) else return (ff, is_unused)
+| none       := do
+  t ← target,
+  is_unused ← recurse_on_expr none ll t,
+  tn ← target,
+  if (t = tn) then return (tt, is_unused) else return (ff, is_unused)
+
+/-- `move_add_arg` is a single elementary argument that `move_add` takes for the
+variables to be moved.  It is either a `pexpr`, or a `pexpr` preceded by a `←`. -/
+meta def move_add_arg (prec : nat) : parser (bool × pexpr) :=
+prod.mk <$> (option.is_some <$> (tk "<-")?) <*> parser.pexpr prec
+
+/-- `move_pexpr_list_or_texpr` is either a list of `move_add_arg`, possibly empty, or a single
+`move_add_arg`. -/
+meta def move_pexpr_list_or_texpr : parser (list (bool × pexpr)) :=
+list_of (move_add_arg 0) <|> list.ret <$> (move_add_arg tac_rbp) <|> return []
+
+/--  Out of a list of `option name`, returns a list of `name`s of target, discarding, if present
+`none`, which corresponds to the goal. -/
+meta def to_hyps : list (option name) → tactic (list expr)
+| []           := pure []
+| (some n::ns) := do ln ← get_local n, fina ← to_hyps ns, return (ln::fina)
+| (none::ns)   := to_hyps ns
+
+/--
+Calling `move_add [a, ← b, c, ← d]`, recursively looks inside the goal for
+expressions involving a sum.  Whenever `move_add` finds a sum, it sorts its terms using
+the user input `[a, ← b, c, ← d]`.  With this input, the output would be
+```lean
+b + d + (sum of terms sorted using the given relation) + a + c.
+```
+Finally, `move_add` can also target hypotheses. If `hp` is in the local context,
+`move_add [← f, g] at hp` performs the rearranging at `hp`.
+-/
```

Comment:
Remember that this docstring is what appears in the tactic documentation page, not the one at the top of the file. It might make sense to shift some text accordingly.

Reply:
Thanks, Eric!  I copied most of the doc-module to the doc-string of the tactic.

Reply:
I would lean towards not duplicating more than a sentence or two of docstring - instead, just tell the reader to read the docstring of `tactic.interactive.move_add` for more information.

Otherwise there's a large risk of the documentation becoming out of sync and inaccurate.